internal/db/redisadapters.go

```diff
@@ -0,0 +1,504 @@
+	}
+
+	return r.rdb.HSet(
+		ctx,
+		keyPrefix+token.ID,
+		r.serializeStruct(encToken)...,
+	).Err()
+}
+
+// SetAccessToken writes the associated ID, access token value, expiration, tokenID and refresh URL
+// of an access token to Redis.
+func (r RedisAdapter) SetAccessToken(ctx context.Context, accessToken models.OauthToken) error {
+	if accessToken.Type != models.AccessTokenType {
+		return fmt.Errorf("token is not of the right type")
+	}
+	return r.setOauthToken(ctx, accessToken)
+}
+
+// SetRefreshToken writes the associated ID, access token value, expiration and tokenID of a refresh token to Redis
+func (r RedisAdapter) SetRefreshToken(ctx context.Context, refreshToken models.OauthToken) error {
+	if refreshToken.Type != models.RefreshTokenType {
+		return fmt.Errorf("token is not of the right type")
+	}
+	return r.setOauthToken(ctx, refreshToken)
+}
+
+func (r RedisAdapter) SetIDToken(ctx context.Context, idToken models.OauthToken) error {
+	if idToken.Type != models.IDTokenType {
+		return fmt.Errorf("token is not of the right type")
+	}
+	return r.setOauthToken(ctx, idToken)
+}
+
+// SetToIndexExpiringTokens writes the associated expiration and tokenID of an access token to Redis
+func (r RedisAdapter) setToIndexExpiringTokens(ctx context.Context, token models.OauthToken) error {
+	var z1 redis.Z
+	z1.Score = float64(token.ExpiresAt.Unix())
+	z1.Member = prefixForTokenType(token.Type) + token.ID
+
+	return r.rdb.ZAdd(
+		ctx,
+		indexExpiringTokens,
+		z1,
+	).Err()
+}
+
+// SetProjectToken writes the project ID and associated expiration and tokenID of a project to Redis
+func (r RedisAdapter) SetProjectToken(ctx context.Context, projectID int, accessToken models.OauthToken) error {
+	z1 := redis.Z{
+		Score:  float64(accessToken.ExpiresAt.Unix()),
+		Member: accessToken.ID,
+	}
+
+	return r.rdb.ZAdd(
+		ctx,
+		projectTokenPrefix+strconv.Itoa(projectID),
+		z1,
+	).Err()
+}
+
+// Remove/delete functions
+
+// RemoveSession removes a session entry from Redis
+func (r RedisAdapter) RemoveSession(ctx context.Context, sessionID string) error {
+	return r.rdb.Del(
+		ctx,
+		sessionPrefix+sessionID,
+	).Err()
+}
+
+// RemoveAccessToken removes an access token entry from Redis
+func (r RedisAdapter) RemoveAccessToken(ctx context.Context, accessToken models.OauthToken) error {
+	err := r.removeFromIndexExpiringTokens(ctx, accessToken)
+	if err != nil {
+		return err
+	}
+
+	return r.rdb.Del(
+		ctx,
+		accessTokenPrefix+accessToken.ID,
+	).Err()
+}
+
+// RemoveRefreshToken removes an access token entry from Redis
+func (r RedisAdapter) RemoveRefreshToken(ctx context.Context, refreshTokenID string) error {
+	return r.rdb.Del(
+		ctx,
+		refreshTokenPrefix+refreshTokenID,
+	).Err()
+}
+
+func (r RedisAdapter) RemoveIDToken(ctx context.Context, idToken models.OauthToken) error {
+	err := r.removeFromIndexExpiringTokens(ctx, idToken)
+	if err != nil {
+		return err
+	}
+
+	return r.rdb.Del(
+		ctx,
+		idTokenPrefix+idToken.ID,
+	).Err()
```

Comment:
this could leak tokens in the db if the second call fails but the first one succeded

Reply:
So this should not pose a problem if id tokens start accumulating in the database. And if they have to be overwritten then it is also not a problem.

The ideal way to address this is to add `TTL` for the token in redis. And redis will then automatically clean tokens up when they are expired. I think we should do this in a followup PR.